picard/acoustid/recordings.py

```diff
@@ -0,0 +1,174 @@
+
+
+class RecordingResolver:
+    """Given an AcoustID lookup result returns a list of MB recordings.
+    The recordings are either directly taken from the AcoustID result or, if the
+    results return only the MBID without metadata, loaded via the MB web service.
+    """
+
+    _recording_map: Dict[str, Dict[str, Recording]]
+
+    def __init__(self, ws: WebService, doc: dict, callback: callable) -> None:
+        self._mbapi = MBAPIHelper(ws)
+        self._doc = doc
+        self._callback = callback
+        self._recording_map = defaultdict(dict)
+        self._recording_cache = dict()
+        self._missing_metadata = deque()
+
+    def resolve(self) -> None:
+        results = self._doc.get('results') or []
+        for result in results:
+            recordings = result.get('recordings') or []
+            result_score = get_score(result)
+            acoustid = result.get('id')
+            for recording in recordings:
+                mbid = recording.get('id')
+                sources = recording.get('sources', 1)
+                if recording_has_metadata(recording):
+                    mb_recording = parse_recording(recording)
+                    self._recording_cache[mbid] = mb_recording
+                    self._recording_map[acoustid][recording['id']] = Recording(
+                        recording=mb_recording,
+                        result_score=result_score,
+                        sources=sources,
+                    )
+                else:
+                    self._missing_metadata.append(IncompleteRecording(
+                        mbid=mbid,
+                        acoustid=acoustid,
+                        result_score=result_score,
+                        sources=sources,
+                    ))
+
+        if self._missing_metadata:
+            self._load_recordings()
+        else:
+            self._send_results()
+
+    def _load_recordings(self):
+        if not self._missing_metadata:
+            self._send_results()
+            return
+
+        mbid = self._missing_metadata[0].mbid
+        if mbid in self._recording_cache:
+            mb_recording = self._recording_cache[mbid]
+            self._recording_request_finished(mb_recording, None, None)
+        else:
+            self._mbapi.get_track_by_id(
+                self._missing_metadata[0].mbid,
+                self._recording_request_finished,
+                inc=['artists', 'release-groups', 'releases', 'media'],
+            )
+
+    def _recording_request_finished(self, mb_recording, http, error):
+        recording = self._missing_metadata.popleft()
+        if error:
+            if error == QNetworkReply.NetworkError.ContentNotFoundError:
+                # Recording does not exist, ignore and move on
+                self._load_recordings()
+            else:
+                self._send_results(error)
+            return
+
+        mbid = mb_recording.get('id')
+        self._recording_cache[mbid] = mb_recording
```

Comment:
Do we still want to use `mbid` here if it's empty? It is tested just below (`if mbid:`)

Reply:
I moved it down into the if